virtual_eeprom/virtual_eeprom.c

```diff
@@ -0,0 +1,219 @@
+};
+
+// Function prototypes
+static int save_eeprom_to_file(struct virtual_eeprom *eeprom);
+static int load_eeprom_from_file(struct virtual_eeprom *eeprom);
+
+// Initialize global list to track all EEPROMs
+static LIST_HEAD(eeprom_list);
+
+static int eeprom_read(void *context, unsigned int offset, void *val, size_t bytes)
+{
+    struct virtual_eeprom *eeprom = context;
+
+    if (offset >= eeprom->size)
+        return -EINVAL;
+
+    if (offset + bytes > eeprom->size)
+        bytes = eeprom->size - offset;
+
+    mutex_lock(&eeprom->mutex);
+    memcpy(val, eeprom->data + offset, bytes);
+    mutex_unlock(&eeprom->mutex);
+
+    return 0;
+}
+
+static int eeprom_write(void *context, unsigned int offset, void *val, size_t bytes)
+{
+    struct virtual_eeprom *eeprom = context;
+    int ret;
+
+    if (offset >= eeprom->size)
+        return -EINVAL;
+
+    if (offset + bytes > eeprom->size)
+        bytes = eeprom->size - offset;
+
+    mutex_lock(&eeprom->mutex);
+    memcpy(eeprom->data + offset, val, bytes);
+    // Save the data to the file after each write
+    ret = save_eeprom_to_file(eeprom);
+    mutex_unlock(&eeprom->mutex);
+
+    return ret < 0 ? ret : 0;
+}
+
+static int save_eeprom_to_file(struct virtual_eeprom *eeprom)
+{
+    struct file *file;
+    loff_t pos = 0;
+    ssize_t ret;
+
+    file = filp_open(eeprom->bin_path, O_WRONLY | O_CREAT | O_TRUNC, 0644);
+    if (IS_ERR(file)) {
+        pr_err("Failed to open file %s for writing: %ld\n", eeprom->bin_path, PTR_ERR(file));
+        return PTR_ERR(file);
+    }
+
+    ret = kernel_write(file, eeprom->data, eeprom->size, &pos);
+    filp_close(file, NULL);
+
+    return ret < 0 ? ret : 0;
+}
+
+static int load_eeprom_from_file(struct virtual_eeprom *eeprom)
+{
+    struct file *file;
+    loff_t pos = 0;
+    ssize_t ret;
+
+    file = filp_open(eeprom->bin_path, O_RDONLY, 0644);
+    if (IS_ERR(file)) {
+	    // don't print read error if the file doesn't exist
+        if (PTR_ERR(file) != -ENOENT)
+            pr_err("Failed to open file %s for reading: %ld\n", eeprom->bin_path, PTR_ERR(file));
+        return PTR_ERR(file);
+    }
+
+    ret = kernel_read(file, eeprom->data, eeprom->size, &pos);
+    filp_close(file, NULL);
+
+    return ret < 0 ? ret : 0;
+}
+
+static int __init eeprom_init(void)
+{
+    struct device_node *np, *child;
+    int ret;
+
+    // Find the parent device node
+    np = of_find_node_by_name(NULL, "virtual_eeproms");
+    if (!np) {
+        pr_err("No device node found for virtual_eeproms\n");
+        return -ENODEV;
+    }
+
+    // Iterate through all child nodes
+    for_each_child_of_node(np, child) {
+        const char *name = child->name;
+        const char *bin_path;
+        u32 size;
+
+        if (of_property_read_u32(child, "size", &size)) {
+            pr_err("Failed to read size for %s\n", name);
+            continue;
+        }
+
+        if (of_property_read_string(child, "bin-path", &bin_path)) {
```

Comment:
It is not a good idea to specify filenames in the device tree.
These should be in module parameters in the driver with default values that can be changed in modprobe/driver config

Reply:
Why not? The location of the backing file of a virtual device seems like the logical counterpart to things like address on the bus etc. so I would have thought the device tree is a perfect fit? Ideally you could pass it into modprobe e.g. for testing, in addition to the device tree

Reply:
Device tree is meant for hardware configuration, not software/files. File locations are usually handled in user space.

Reply:
> Device tree is meant for hardware configuration, not software/files. File locations are usually handled in user space.

Agree, but here we're adding something that pretends to be hardware?

Reply:
I think it's more similar to a bluetooth module where the hardware is specified in the device tree but the firmware image that gets loaded onto the bluetooth device is on the file system. It would never be put in the device tree. I'm happy to defer if you want though. Can sonarqube settle it?

Reply:
I'm not too fussed. As long as it's not baked into the kernel module I'm happy. Having it in the modprobe conf just means there's yet another file involved, but I believe Yocto deals with this reasonably well, i.e. you don't have to manually create and add a file, I think?

Reply:
Ah, and sonarqube won't settle this as it doesn't look at dts files

Reply:
Removed path config from dts and defined default directory in module which is configurable with parameter while loading. The module appends the file names to this directory so we only have to pass one parameter.